src/wifi.cpp

```diff
@@ -0,0 +1,694 @@
+// Resources:
+// - Create a basic POSIX TPC/IP server in C:
+//   https://medium.com/@coderx_15963/basic-tcp-ip-networking-in-c-using-posix-9a074d65bb35
+// - How to connect your ESP32 to Wifi and an existing POSIX TCP/IP server:
+//   https://www.youtube.com/watch?v=_dRrarmQiAM
+//   https://github.com/espressif/esp-idf/blob/v5.3.1/examples/wifi/getting_started/station/main/station_example_main.c
+//   https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_netif.html
+//   https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/esp_event.html
+//   https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_wifi.html
+
+#include "wifi.h"
+
+#if WIFI_ENABLED
+
+// Include FreeRTOS common header
+#include "freertos/FreeRTOS.h"
+// Include FreeRTOS task API
+#include "freertos/task.h"
+// Include FreeRTOS event group API
+#include "freertos/event_groups.h"
+// Import light-weight IP socket API
+#include "lwip/sockets.h"
+// Include ESP WiFi API
+#include "esp_wifi.h"
+// Include ESP event API
+#include "esp_event.h"
+
+// Include custom Menu class implementation
+#include "menu.h"
+
+// ====================================== //
+// Define useful constants and data types //
+// ====================================== //
+
+// Define bitmask for success in WiFi event loop
+#define WIFI_CONNECTED_BIT BIT0
+// Define bitmask for failure in WiFi event loop
+#define WIFI_FAIL_BIT BIT1
+// Define the max number of retries allowed to try to connect to one AP in a row
+#define NUM_MAX_WIFI_CONNECT_RETRIES 5
+
+// Define the number of currently supported TCP commands
+#define NUM_TCP_COMMANDS 3
+
+// Define, when receiving a TCP packet, what special strings should cause what actions
+typedef struct tcp_command_s {
+    // The string, that if the TCP packet matches, should trigger an action
+    const char* command;
+    // The action to be triggered if the TCP packet matched command
+    void (*action)();
+} tcp_command_t;
+
+// ======================= //
+// Instantiate useful data //
+// ======================= //
+
+// Keep track of how many times the device has retried connecting to one AP in a row
+size_t num_wifi_connect_retries = 0;
+// Keep track of the handle to the WiFi event group, so we can set success and failure bits on its status
+EventGroupHandle_t event_group_handle_wifi = nullptr;
+// Keep track of the file handle being used as the IP socket connection
+// TODO: Put a mutex around this file handle
```

Comment:
Do this.

Reply:
Coding to account for this adds a lot of code size and complexity. I'll address this if it ever becomes and issue instead of making the code perfect right now, leaving notes.